src/runtime/virtcontainers/fs_share_linux.go

```diff
@@ -507,3 +557,225 @@ func (f *FilesystemShare) UnshareRootFilesystem(ctx context.Context, c *Containe
 	return nil
 
 }
+
+func (f *FilesystemShare) watchDir(source string) error {
+
+	// Add a watcher for the configmap, secrets, projected-volumes and downwar-api directories
+	// /var/lib/kubelet/pods/<uid>/volumes/{kubernetes.io~configmap, kubernetes.io~secrets, kubernetes.io~downward-api, kubernetes.io~projected-volume}
+
+	// Note: From fsNotify docs - https://pkg.go.dev/github.com/fsnotify/fsnotify
+	// Watching individual files (rather than directories) is generally not
+	// recommended as many tools update files atomically. Instead of "just"
+	// writing to the file a temporary file will be written to first, and if
+	// successful the temporary file is moved to to destination removing the
+	// original, or some variant thereof. The watcher on the original file is
+	// now lost, as it no longer exists.
+	// Instead, watch the parent directory and use Event.Name to filter out files
+	// you're not interested in.
+
+	// Also fsNotify doesn't add watcher recursively. So we need to walk the root directory and add the required watches
+
+	f.Logger().Infof("watchDir: Add fsnotify watcher for dir (%s)", source)
+	watchList := f.watcher.WatchList()
+	if len(watchList) == 0 {
+		f.Logger().Infof("watchDir: Adding fsnotify watcher for dir (%s)", source)
+		err := f.watcher.Add(source)
+		if err != nil {
+			f.Logger().Info("watchDir: Failed to add watcher to list")
+			return err
+		}
+	}
+
+	for _, v := range watchList {
+		if v == source {
+			f.Logger().Infof("watchDir: Watcher for dir(%s) is already present", source)
+			return nil
+		}
+	}
+	f.Logger().Infof("watchDir: Adding fsnotify watcher for dir (%s)", source)
+	err := f.watcher.Add(source)
+	if err != nil {
+		f.Logger().Info("watchDir: Failed to add watcher to list")
+		return err
+	}
+
+	return nil
+
+}
+
+func (f *FilesystemShare) StartFileEventWatcher(ctx context.Context) error {
+
+	// Start event loop if watchList is not empty
+	if (f.watcher != nil) || len(f.watcher.WatchList()) == 0 {
+		f.Logger().Info("StartFileEventWatcher: No watches found, returning")
+		return nil
+	}
```

Comment:
```
f.watcher != nil
```
I think this should be
```
f.watcher == nil
```
Otherwise, `StartFileEventWatcher` immediately returns in my setup. 

Alternatively:
```
	if !(f.watcher != nil && len(f.watcher.WatchList() > 0) {
```

Reply:
Nice catch. My bad. I'll fix it in the next iteration